### PR TITLE
Backport bitcoin#10637 (partial)

### PR DESCRIPTION
**Pull request by UdjinM6**
Backports bitcoin#10637

This one is pretty complex, backported it commit-by-commit. It's partial because I omitted most of the changes in `CWallet::CreateTransaction` (see 5137de5, original commit https://github.com/bitcoin/bitcoin/pull/10637/commits/6a34ff5335786615771ca423134a484b04831c4e), so bnb is not actually used there atm and the main reason to backport it is simply to avoid future merge conflicts everywhere else.

### CODE REVIEW

File: src/script/sign.cpp [hunk 1]
```diff
@@ -158,7 +158,12 @@ SignatureData DataFromTransaction(const CMutableTransaction& tx, unsigned int nI
...
 {
     assert(tx.vin.size() > nIn);
-    tx.vin[nIn].scriptSig = data.scriptSig;
+    UpdateInput(tx.vin[nIn], data);
+}
+
+void UpdateInput(CTxIn& input, const SignatureData& data)
+{
+    input.scriptSig = data.scriptSig;
 }
```
Comment (PastaPastaPasta) [inline src/script/sign.cpp:167]: Shouldn't UpdateInput be above UpdateTransaction?

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
shouldn't f9509816be8dffdbfba85640e8108e4255b8b7a4 be called "More of 10637..."?

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Review by xdustinface — APPROVED** [review]
utACK

**Pull request merged**